File: gitdot-api/src/endpoint/repository/get_repository_commits.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::default_ref;
use crate::{endpoint::Endpoint, resource::repository::RepositoryCommitsResource};

pub struct GetRepositoryCommits;

impl Endpoint for GetRepositoryCommits {
    const PATH: &'static str = "/repository/{owner}/{repo}/commits";
    const METHOD: http::Method = http::Method::GET;

    type Request = GetRepositoryCommitsRequest;
    type Response = GetRepositoryCommitsResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct GetRepositoryCommitsRequest {
    #[serde(default = "default_ref")]
    pub ref_name: String,
    pub from: Option<DateTime<Utc>>,
    pub to: Option<DateTime<Utc>>,
}

pub type GetRepositoryCommitsResponse = RepositoryCommitsResource;

File: gitdot-web/packages/api/src/endpoint/repository/get-repository-commits.ts
import { z } from "zod";
import { RepositoryCommitsResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const GetRepositoryCommitsRequest = z.object({
  ref_name: z.string().optional(),
  from: z.iso.datetime().optional(),
  to: z.iso.datetime().optional(),
});
export type GetRepositoryCommitsRequest = z.infer<
  typeof GetRepositoryCommitsRequest
>;

export const GetRepositoryCommitsResponse = RepositoryCommitsResource;
export type GetRepositoryCommitsResponse = z.infer<
  typeof GetRepositoryCommitsResponse
>;

export const GetRepositoryCommits = {
  path: "/repository/{owner}/{repo}/commits",
  method: "GET",
  request: GetRepositoryCommitsRequest,
  response: GetRepositoryCommitsResponse,
} as const satisfies Endpoint;
export type GetRepositoryCommits = typeof GetRepositoryCommits;
